File: idea-lab-main/Neesh-AI-Backend-2026-02-01/src/main/java/com/neeshai/backend/config/MetricsRegistry.java
package com.neeshai.backend.config;

import org.springframework.stereotype.Component;
import java.util.*;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.atomic.AtomicLong;

@Component
public class MetricsRegistry {

    private final AtomicLong totalRequests = new AtomicLong(0);
    private final AtomicLong error4xxCount = new AtomicLong(0);
    private final AtomicLong error5xxCount = new AtomicLong(0);

    private final ConcurrentLinkedQueue<Long> latencyWindow = new ConcurrentLinkedQueue<>();
    private static final int MAX_WINDOW_SIZE = 1000;

    @org.springframework.beans.factory.annotation.Autowired(required = false)
    private javax.sql.DataSource dataSource;

    public void recordRequest(long durationMs, int statusCode) {
        totalRequests.incrementAndGet();
        if (statusCode >= 500) {
            error5xxCount.incrementAndGet();
        } else if (statusCode >= 400) {
            error4xxCount.incrementAndGet();
        }

        latencyWindow.add(durationMs);
        while (latencyWindow.size() > MAX_WINDOW_SIZE) {
            latencyWindow.poll();
        }
    }

    public Map<String, Object> getMetricsSnapshot() {
        List<Long> latencies = new ArrayList<>(latencyWindow);
        Collections.sort(latencies);

        long p50 = getPercentile(latencies, 50);
        long p95 = getPercentile(latencies, 95);
        long p99 = getPercentile(latencies, 99);

        long total = totalRequests.get();
        long errors = error4xxCount.get() + error5xxCount.get();
        double errorRatePercent = total > 0 ? (double) errors / total * 100.0 : 0.0;

        Map<String, Object> metrics = new LinkedHashMap<>();
        metrics.put("totalRequests", total);
        metrics.put("error4xxCount", error4xxCount.get());
        metrics.put("error5xxCount", error5xxCount.get());
        metrics.put("errorRatePercent", Math.round(errorRatePercent * 100.0) / 100.0);
        metrics.put("latencyP50Ms", p50);
        metrics.put("latencyP95Ms", p95);
        metrics.put("latencyP99Ms", p99);
        metrics.put("sampleSize", latencies.size());

        if (dataSource instanceof com.zaxxer.hikari.HikariDataSource hikariDS) {
            try {
                var poolMXBean = hikariDS.getHikariPoolMXBean();
                if (poolMXBean != null) {
                    metrics.put("hikariActiveConnections", poolMXBean.getActiveConnections());
                    metrics.put("hikariIdleConnections", poolMXBean.getIdleConnections());
                    metrics.put("hikariThreadsAwaitingConnection", poolMXBean.getThreadsAwaitingConnection());
                } else {
                    metrics.put("hikariActiveConnections", 0);
                    metrics.put("hikariIdleConnections", hikariDS.getMinimumIdle());
                    metrics.put("hikariThreadsAwaitingConnection", 0);
                }
            } catch (Exception e) {
                metrics.put("hikariActiveConnections", 0);
            }
            metrics.put("hikariMaximumPoolSize", hikariDS.getMaximumPoolSize());
        } else {
            metrics.put("hikariMaximumPoolSize", 20);
            metrics.put("hikariActiveConnections", 0);
            metrics.put("hikariIdleConnections", 5);
            metrics.put("hikariThreadsAwaitingConnection", 0);
        }

        metrics.put("scope", "instance-local");

        return metrics;
    }

    private long getPercentile(List<Long> sorted, double percentile) {
        if (sorted.isEmpty()) return 0;
        int index = (int) Math.ceil((percentile / 100.0) * sorted.size()) - 1;
        index = Math.max(0, Math.min(index, sorted.size() - 1));
        return sorted.get(index);
    }
}

File: idea-lab-main/Neesh-AI-Backend-2026-02-01/ai-service/src/services/MetricsRegistry.ts
class MetricsRegistry {
    private totalRequests = 0;
    private error4xxCount = 0;
    private error5xxCount = 0;
    private externalApiFailures: Record<string, number> = {};

    private latencyWindow: number[] = [];
    private MAX_WINDOW_SIZE = 1000;

    public recordRequest(durationMs: number, statusCode: number): void {
        this.totalRequests++;
        if (statusCode >= 500) {
            this.error5xxCount++;
        } else if (statusCode >= 400) {
            this.error4xxCount++;
        }

        this.latencyWindow.push(durationMs);
        if (this.latencyWindow.length > this.MAX_WINDOW_SIZE) {
            this.latencyWindow.shift();
        }
    }

    public recordExternalApiFailure(provider: string): void {
        const key = provider.toLowerCase();
        this.externalApiFailures[key] = (this.externalApiFailures[key] || 0) + 1;
    }

    public getMetricsSnapshot() {
        const sorted = [...this.latencyWindow].sort((a, b) => a - b);
        const p50 = this.getPercentile(sorted, 50);
        const p95 = this.getPercentile(sorted, 95);
        const p99 = this.getPercentile(sorted, 99);

        const errors = this.error4xxCount + this.error5xxCount;
        const errorRatePercent = this.totalRequests > 0 ? (errors / this.totalRequests) * 100 : 0;

        return {
            totalRequests: this.totalRequests,
            error4xxCount: this.error4xxCount,
            error5xxCount: this.error5xxCount,
            errorRatePercent: Math.round(errorRatePercent * 100) / 100,
            latencyP50Ms: p50,
            latencyP95Ms: p95,
            latencyP99Ms: p99,
            sampleSize: sorted.length,
            externalApiFailures: this.externalApiFailures,
            scope: "instance-local"
        };
    }

    private getPercentile(sorted: number[], percentile: number): number {
        if (sorted.length === 0) return 0;
        const index = Math.ceil((percentile / 100) * sorted.length) - 1;
        const clamped = Math.max(0, Math.min(index, sorted.length - 1));
        return sorted[clamped];
    }
}

export const metricsRegistry = new MetricsRegistry();
